app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayController.kt

```diff
@@ -134,18 +143,37 @@ class DefaultTabsTrayController(
      * Deletes the [TabSessionState] with the specified [tabId].
      *
      * @param tabId The id of the [TabSessionState] to be removed from TabsTray.
+     * @param source app feature from which the tab with [tabId] was closed.
+     * @param cancelPrivateDownloadsAccepted user confirmed private downloads cancellation
      * This method has no effect if the tab does not exist.
      */
-    override fun handleTabDeletion(tabId: String) {
+    override fun handleTabDeletion(
+        tabId: String,
+        source: String?,
+        cancelPrivateDownloadsAccepted: Boolean
+    ) {
         val tab = browserStore.state.findTab(tabId)
 
         tab?.let {
             if (browserStore.state.getNormalOrPrivateTabs(it.content.private).size != 1) {
                 tabsUseCases.removeTab(tabId)
                 showUndoSnackbarForTab(it.content.private)
             } else {
-                dismissTabsTrayAndNavigateHome(tabId)
+                val privateDownloads = browserStore.state.downloads.filter { it.value.private }
+                if (privateDownloads.isNotEmpty() && !cancelPrivateDownloadsAccepted) {
+                    showCancelledDownloadWarning(
+                        CancelDownloadFragmentArguments(
+                            privateDownloads.size,
+                            Trigger.SINGLE_TAB,
+                            tabId = tab.id
+                        )
+                    )
+                    return
+                } else {
+                    dismissTabsTrayAndNavigateHome(tabId)
+                }
             }
+            metrics.track(Event.ClosedExistingTab(source ?: "unknown"))
```

Comment:
I am not quite sure what this metrics does, but it made me go - hm, the reviewer might want to take a closer look.